SyntaxKit/RegularExpression.swift

```diff
@@ -0,0 +1,139 @@
+//
+//  RegularExpression.swift
+//  SyntaxKit
+//
+//  Created by Rachel on 2021/2/19.
+//  Copyright © 2021 Sam Soffes. All rights reserved.
+//
+
+import Foundation
+
+internal class RegularExpression {
+
+    // MARK: - Properties
+
+    var pattern: String { return _pattern }
+    var options: RegularExpression.Options { return _options }
+    var isTemplate: Bool { return _isTemplate }
+    var expression: NSRegularExpression? { return _expression }
+
+    // swiftlint:disable strict_fileprivate
+    fileprivate var _pattern: String
+    fileprivate var _options: RegularExpression.Options
+    fileprivate var _isTemplate: Bool
+    fileprivate var _expression: NSRegularExpression?
+    // swiftlint:enable strict_fileprivate
+
+    // MARK: - Initializer
+
+    init(pattern: String, options: RegularExpression.Options = []) throws {
+        self._pattern = pattern
+        self._options = options
+        self._isTemplate = pattern.hasBackReferencePlaceholder
+        if !self._isTemplate {
+            self._expression = try NSRegularExpression(pattern: pattern, options: NSRegularExpression.Options(rawValue: options.rawValue))
```

Comment:
It would be better to add expandBackReferences to this class and not expose the expression at all. It would also be good to have the methods NSRegularExpression methods below fail if expression is nil.